File: Java/CombinationSum.java
class Solution {
    private void comb(int[] nums, int target, int start, 
                    List<Integer> elem, List<List<Integer>> res) {
        if(target == 0) {
            res.add(new ArrayList<>(elem));
            return;
        }
        if(target < 0) return;

        for(int i = start; i < nums.length; i++) {
            elem.add(nums[i]);
            comb(nums, target - nums[i], i, elem, res);
            elem.remove(elem.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> elem = new ArrayList<>();
        comb(candidates, target, 0, elem, result);
        return result;
    }
}

File: Go/combination_sum.go
func combinationSum(candidates []int, target int) [][]int {
    var res [][]int
    var backtrack func(cur, start int, elem []int)
    backtrack = func(cur, start int, elem []int) {
        if cur == 0 {
            res = append(res, append([]int{}, elem...))
            return
        }
        if cur < 0 {return}
        for i := start; i < len(candidates); i++ {
            elem = append(elem, candidates[i])
            backtrack(cur - candidates[i], i, elem)
            elem = elem[:len(elem) - 1]
        }
    }
    backtrack(target, 0, []int{})
    return res
}